File: LeetCode/Python Solutions/Pow(x, n)/power.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        # Base case: when n is zero, result is always 1 
        if n == 0:
            return 1.0

        # if n is negative, change the problem to x^-n to 1/x^n
        if n < 0:
            return self.myPow(1/x, -n)

        # Recursive case
        else:
            half = self.myPow(x, n // 2)  # Calculate power of half n
            
            # If n is even, the result is half*half
            if n % 2 == 0:
                return half * half
            # If n is odd, the result is half*half*x
            else:
                return half * half * x

File: LeetCode/Java Solutions/Pow(x, n)/power.java
class Solution {
    public double myPow(double x, int n) {
        long N = n;
        // if n is negative, change the problem to x^-n to 1/x^n
        if (N < 0) {
            x = 1 / x;
            N = -N;
        }

        // Initialize answer to 1.0 as any number raised to the power of 0 is 1
        double ans = 1.0;
        
        // current_product represents the current base that is used to compute the power
        double current_product = x;

        // We will now start reducing the power N towards 0
        for (long i = N; i > 0; i /= 2) {
            // If the current power is odd, we multiply the answer with the current base
            if ((i % 2) == 1) {
                ans *= current_product;
            }
            // Now we square the current base for the next iteration
            current_product *= current_product;
        }
        return ans;
    }
}
